README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->
[![Build
Status](https://travis-ci.org/blasee/fetchR.svg)](https://travis-ci.org/blasee/fetchR)
[![CRAN\_Status\_Badge](https://r-pkg.org/badges/version/fetchR)](https://cran.r-project.org/package=fetchR)
[![](https://cranlogs.r-pkg.org/badges/fetchR)](https://cran.r-project.org/package=fetchR)

Wind fetch
==========

Wind fetch is an important measurement in coastal applications. It
provides a measurement for the unobstructed length of water over which
wind from a certain direction can blow over. The higher the wind fetch
from a certain direction, the more energy is imparted onto the surface
of the water resulting in a larger sea state. Therefore, the larger the
fetch, the larger the exposure to wind and the more likely the site
experiences larger sea states.

Why **fetchR**?
===============

Averaging the wind fetch for numerous directions at the same location is
a reasonable measure of the overall wind exposure. This process of
calculating wind fetch can be extremely time-consuming and tedious,
particularly if a large number of fetch vectors are required at many
locations. The **fetchR** package calculates wind fetch for any marine
location on Earth. There are also plot methods to help visualise the
wind exposure at the various locations, and methods to output the fetch
vectors to a KML file for further investigation.

Installation
============

You can install and load the latest version of **fetchR** from CRAN.

``` r
# Install fetchR
install.packages("fetchR")

# Load the fetchR package
library(fetchR)
```

Calculating wind fetch with **fetchR**
======================================

If you already have a `SpatialPolygons` object representing the
coastline and surrounding islands, and a `SpatialPoints` object
representing the locations, then calculating wind fetch with **fetchR**
is easy. You can just pass these two arguments into the `fetch`
function.

![Map showing the projected SpatialPolygons (filled grey polygons) and
SpatialPoints (black crosses) objects.](./vignettes/figures/proj.png)

##### Note

The data for the polygon layer was sourced from [Land Information New
Zealand](https://data.linz.govt.nz/layer/1153-nz-coastlines-and-islands-polygons-topo-150k/)[1].

``` r
# Calculate wind fetch by passing in the projected SpatialPolygons object (nz_poly_proj)
# and the projected SpatialPoints object (fetch_locs_proj) to the fetch function.
my_fetch_proj = fetch(nz_poly_proj, fetch_locs_proj)

my_fetch_proj
```

    #> Is projected : TRUE
    #> Max distance : 300 km
    #> Directions   : 36
    #> Sites        : 3
    #> 
    #>                  North East South West Average
    #> Kawau Bay          2.3  8.6   6.7  4.8     5.6
    #> Chetwode Islands 158.9 75.4  12.1 17.1    65.9
    #> Foveaux Strait    27.6 89.5 112.4 96.8    81.6

The `my_fetch_proj` provides a summary of the fetch for all the four
quadrants, along with an overall average of the fetch length at all the
sites.

Visualise the fetch vectors
===========================

``` r
# Plot the fetch vectors, along with the coastline and surrounding islands
plot(my_fetch_proj, nz_poly_proj)
```

![Fetch vectors at all the sites](./vignettes/figures/nz_fetch.png)

![Fetch vectors at Kawau Bay](./vignettes/figures/kawau.png)

![Fetch vectors at the Chetwode
Islands](./vignettes/figures/chetwode.png)

![Fetch vectors at Foveaux Strait](./vignettes/figures/foveaux.png)

Export to a KML file
====================

``` r
# Export the fetch vectors to a KML file for further investigation
kml(my_fetch_proj, colour = "white")
```

Note that the distances calculated in Google Earth are (almost) the same
as the distances calculated with **fetchR**. This can be seen in the KML
output as the fetch vector at 90 degrees for the Foveaux Strait site is
300km (the maximum distance by default) in both **fetchR** and Google
Earth, although these algorithms differ.

![Output to KML](./vignettes/figures/kml.png)

Get started with **fetchR**
===========================

Read the short introductory vignette to get you started with **fetchR**,
and have a look at the simple, reproducible example in the `fetch`
function.

``` r
# Read the short vignette
vignette("introduction-to-fetchR")

# Reproduce a simple example
example(fetch)
```

Citation
========

``` r
citation("fetchR")
#> 
#> To cite package 'fetchR' in publications use:
#> 
#>   Blake Seers (2020). fetchR: Calculate Wind Fetch. R package version
#>   2.1-2. https://cran.r-project.org/package=fetchR
#> 
#> A BibTeX entry for LaTeX users is
#> 
#>   @Manual{,
#>     title = {fetchR: Calculate Wind Fetch},
#>     author = {Blake Seers},
#>     year = {2020},
#>     note = {R package version 2.1-2},
#>     url = {https://cran.r-project.org/package=fetchR},
#>   }
```

[1] This README contains data that was sourced from Land Information New
Zealand without adaptation that is protected under CC-By Land
Information New Zealand.


NEWS.md
# fetchR 2.1-2

## Bug fixes

* Fix future warnings from CRAN checks on R-devel (#10)

# fetchR 2.1-1

## Minor changes

* You can now specify up to 1 direction per degree on the compass rose 
(`n_directions = 90`).

# fetchR 2.1-0

## Minor changes

* Site names are automatically read from the data associated with the shapefile.
  The names have to be in a column with a name matching the regular expression;
  `^[Nn]ames{0,1}$`.
  
* Created a [shiny application](https://github.com/blasee/fetchR#shiny-application) for **fetchR**.

* Use `on.exit` to close the KML file connection in `kml,Fetch` method.

# fetchR 2.0-0

## Major changes

* **fetchR** is now generalised to incorporate `SpatialPolyons` representing any
marine location on Earth (#6). As a result, **fetchR** 2.0 is no longer limited 
to calculating wind fetch within New Zealand coastal areas.

### Rewrite the entire algorithm for calculating wind fetch. 

* The `rgeos::gBuffer` function is used to calculate the end points of the fetch 
vectors at their maximum distance.

* The interactions between the fetch vectors and the coastlines, or any of the 
polygons, are calculated with the `rgeos::gIntersection` function as opposed to
the slow, iterative algorithm used by the package's predecessors.

* Computation times have dramatically decreased.

## Minor changes

* Create a vignette for the package; `vignette("introduction-to-fetchR")`.


R/fetch.R
#' Calculate Wind Fetch
#' 
#' Wind fetch is the unobstructed length of water over which wind can blow, and
#' it is commonly used as a measure of exposure to wind and waves at coastal 
#' sites. The \code{fetch} function automatically calculates the wind fetch for
#' marine locations within the boundaries of the specified coastline layer.
#' This allows wind fetch to be calculated anywhere around the globe.
#' 
#' The function takes a \code{\link[sp]{SpatialPolygons-class}} object 
#' (\code{polygon_layer}) that represents the coastline, surrounding islands, 
#' and any other obstructions, and calculates the wind fetch for every specified 
#' direction. This is calculated for all the user-defined sites, that are 
#' represented as the point geometries in a 
#' \code{\link[sp]{SpatialPoints-class}} object.
#' 
#' The directions for which the wind fetch are calculated for each site are 
#' determined by the number of directions per quadrant (\code{n_directions}). 
#' The default value of 9 calculates 9 fetch vectors per quadrant (90 degrees), 
#' or equivalently, one fetch vector every 10 degrees. The first fetch vector is 
#' always calculated for the northerly direction (0/360 degrees).
#' 
#' @param polygon_layer \code{\link[sp]{SpatialPolygons}}* object where the
#'                      polygon geometries represent any obstructions to fetch
#'                      calculations including the coastline, islands and/or 
#'                      exposed reefs.
#' @param site_layer \code{\link[sp]{SpatialPoints}}* object where the point 
#'                   geometries represent the site locations.
#' @param max_dist numeric. Maximum distance in kilometres (default 300). This 
#'                 will need to be scaled manually if the units for the CRS are 
#'                 not 'm'.
#' @param n_directions numeric. The number of fetch vectors to calculate per 
#'                     quadrant (default 9).
#' @param site_names character vector of the site names. If missing, the site 
#'                   names are taken from a column of the data associated with 
#'                   \code{site_layer} matching the regular expression 
#'                   \code{^[Nn]ames{0,1}}. If there is no such column, then 
#'                   default names are created ('Site 1', 'Site 2', ...).
#' @param quiet logical. Suppress diagnostic messages? (Default \code{FALSE}).
#' 
#' @return Returns a \code{\link{Fetch}} object.
#' 
#' @note At least one of the inputs to the \code{polygon_layer} or 
#'       \code{site_layer} arguments must be projected. If one of the inputs are 
#'       not projected, then it will be transformed to have the same projection 
#'       as the other. If both are projected, but do not have identical 
#'       coordinate reference systems (CRS) then \code{site_layer} will be 
#'       transformed to the same CRS as \code{polygon_layer}.
#' 
#' @seealso \code{\link[rgdal]{spTransform}} for methods on transforming map 
#'          projections and datum.
#' @seealso \code{\link[sp]{is.projected}} for checking whether a spatial object 
#'          is projected.
#' @seealso \code{\link{fetchR}} for an overview of this package with an
#'          extensive, reproducible example.
#' @seealso \code{\link{summary,Fetch-method}} for summarising the fetch lengths.
#' 
#' @importFrom rgdal CRSargs
#' @importFrom sp SpatialPoints CRS over spTransform coordinates SpatialLinesLengths SpatialLinesDataFrame identicalCRS
#' @importFrom rgeos gBuffer gIntersects gIntersection
#' @importFrom utils head
#' @importFrom methods new is
#' @import sp
#' @examples
#' 
#' # Create the polygon layer ----------------------------------------
#' #
#' # This is the layer that represents any obstacles that obstruct wind flow.
#' 
#' # Import map data for the Philippines.
#' philippines.df = ggplot2::map_data("world", region = "Philippines")
#' 
#' # Create a list for each separate polygon
#' philippines.list = split(philippines.df[, c("long", "lat")], 
#'                          philippines.df$group)
#'                          
#' library(sp)
#' 
#' philippines.Poly = lapply(philippines.list, Polygon)
#' philippines.Polys = list(Polygons(philippines.Poly, ID = "Philippines"))
#' 
#' # Include CRS information to make it a SpatialPolygons object
#' philippines.sp = SpatialPolygons(philippines.Polys, 
#'                                      proj4string = CRS("+init=epsg:4326"))
#' 
#' # Create the points layer ----------------------------------------
#' #
#' # The points layer represents the locations for which the wind fetch needs to
#' # be calculated.
#' 
#' # We need to calculate wind fetch for the following 3 sites:
#' sites.df = data.frame(lon = c(124.4824, 125.8473, 124.8416),
#'                       lat = c(9.167999, 9.751394, 11.478243),
#'                       site = c("Camiguin Island", "Bucas Grande Island",
#'                                "Talalora"))
#'                       
#' # Create the SpatialPoints object
#' sites.sp = SpatialPoints(sites.df[, 1:2], CRS("+init=epsg:4326"))
#' 
#' # Map projection -------------------------------------------------
#' #
#' # At least one of the polygon or points layers need to be projected to 
#' # calculate wind fetch.
#' 
#' # All these locations lie within the Philippines zone 5 / PRS92, that has
#' # WGS84 Bounds: 123.8000, 5.3000, 126.7000, 12.7500
#' # (http://spatialreference.org/ref/epsg/3125/)
#' # This suggests that this is a suitable map projection.
#' philippines.proj = spTransform(philippines.sp, "+init=epsg:3125")
#' 
#' # Calculate wind fetch -------------------------------------------
#' # 
#' # Calculate wind fetch at all the 3 locations for every 10 degrees on the
#' # compass rose, with a maximum distance for any fetch vector of 300 km.
#' my_fetch = fetch(philippines.proj, sites.sp, site_names = sites.df$site)
#' my_fetch
#' 
#' # Return only the summary data frame
#' summary(my_fetch)
#' 
#' # Transform the fetch vectors back to the original CRS
#' my_fetch_latlon = spTransform(my_fetch, proj4string(philippines.sp))
#' 
#' # Return the raw data in the original, lat/lon coordinates
#' my_fetch_latlon.df = as(my_fetch_latlon, "data.frame")
#' my_fetch_latlon.df
#' 
#' # Plot the wind fetch vectors ------------------------------------
#' 
#' # Plot the fetch vectors in the projected space...
#' plot(my_fetch, philippines.proj, axes = TRUE)
#' 
#' # ... or in the original coordinate reference system
#' plot(my_fetch, philippines.sp, axes = TRUE)
#' 
#' # Output to KML --------------------------------------------------
#' \dontrun{
#' 
#' # Save a KML file in the current working directory.
#' kml(my_fetch, colour = "white")
#' }
#' @export
fetch = function(polygon_layer, site_layer, max_dist = 300, n_directions = 9,
                 site_names, quiet = FALSE){
  
  if (!is(polygon_layer, "SpatialPolygons"))
    stop(paste("polygon_layer must be a SpatialPolygons object.\nSee",
               "'?SpatialPolygons' for details on how to create a",
               "SpatialPolygons object."), call. = FALSE)
  
  if (!is(site_layer, "SpatialPoints"))
    stop(paste("site_layer must be a SpatialPoints object.\nSee",
         "'?SpatialPoints' for details on how to create a SpatialPoints",
         "object."), call. = FALSE)
  
  if (!is.numeric(max_dist) || length(max_dist) != 1)
    stop("max_dist must be a single number.", call. = FALSE)
  
  if (!is.numeric(n_directions) || length(n_directions) != 1)
    stop("n_directions must be a single integer.", call. = FALSE)
  n_directions = round(n_directions)
  
  if (n_directions < 1 || n_directions > 90)
    stop("n_directions must be between 1 and 90.", call. = FALSE)
  
  if (!missing(site_names)){
    site_names = as.character(site_names)
    
    if (length(site_names) != length(site_layer)){
      warning(paste("lengths differ for the number of sites and site names;", 
                    "using default names instead."), call. = FALSE)
      site_names = paste("Site", seq_along(site_layer))
    }
  } else {
    if (any(grepl("^[Nn]ames{0,1}$", names(site_layer)))){
      name_col = grep("^[Nn]ames{0,1}$", names(site_layer))
      site_names = as.character(data.frame(site_layer)[, name_col[[1]]])
    } else {
      site_names = paste("Site", seq_along(site_layer))
    }
  }
  
  quiet = as.logical(quiet[1])
  
  ## Check if the polygon and points layers are projected, and ensure they have 
  ## the same CRS.
  which_proj = c(is.projected(polygon_layer), is.projected(site_layer))
  
  if (all(!which_proj))
    stop("polygon_layer and/or site_layer must be projected to calculate fetch", 
         call. = FALSE)
  
  if (all(which_proj) && !identicalCRS(polygon_layer, site_layer)){
    warning("the CRS for polygon_layer and site_layer differ; transforming
              site_layer CRS to match")
    site_layer = spTransform(site_layer, CRS(proj4string(polygon_layer)))
  }
  
  if (!which_proj[1]){
    if (!quiet)
      message("projecting polygon_layer onto the site_layer CRS")
    polygon_layer = spTransform(polygon_layer, CRS(proj4string(site_layer)))
  }
  
  if (!which_proj[2]){
    if (!quiet)
      message("projecting site_layer onto the polygon_layer CRS")
    site_layer = spTransform(site_layer, CRS(proj4string(polygon_layer)))
  }
  
  if (!quiet)
    message("checking site locations are not on land")
  
  # Should remove these sites with warning instead of returning error.
  if (any(!is.na(over(polygon_layer, site_layer))))
    stop("at least one site location is on land")
  
  # Convert max_dist to appropriate units.
  # First of all convert max_dist to metres (default)
  max_dist = max_dist * 1000
  
  # Double check if metres are the correct units
  proj_unit = strsplit(gsub("*.+units=", "", 
                            CRSargs(CRS(proj4string(polygon_layer)))), " ")[[1]][1]
  
  # If not, warn the user that the supplied max_dist should be scaled 
  # appropriately
  if (proj_unit != "m")
    warning("the PROJ.4 unit is not metres; ensure max_dist has been scaled 
            appropriately")
  
  directions = head(seq(0, 360, by = 360 / (n_directions * 4)), -1)
  
  # Return the quadrant the directions belong to
  dirs = as.numeric(directions)
  dirs_bin = findInterval(dirs, seq(45, 315, by = 90))
  quadrant = rep("North", length(dirs))
  quadrant[dirs_bin == 1] = "East"
  quadrant[dirs_bin == 2] = "South"
  quadrant[dirs_bin == 3] = "West"
  
  # Rearrange sequence order to start at 90 degrees to match up with the output
  # from gBuffer
  directions = unlist(split(directions, directions < 90), use.names = FALSE)
  
  # Create an empty list to store the Fetch objects
  fetch_list = vector("list", length(site_layer))
  
  for (i in seq_along(site_layer)){
    
    message("calculating fetch for ", site_names[i], " (", i, " out of ", 
            length(site_layer), ")")
    
    ## Create polygon (approximating a circle) with a given radius. These 
    ## vertices are used for creating the end points for the fetch vectors.
    d_bff = gBuffer(site_layer[i, ], width = max_dist, quadsegs = n_directions)
    
    ## Calculate end points at the maximum distances.
    fetch_ends = head(coordinates(d_bff@polygons[[1]]@Polygons[[1]]), -1)
    fetch_ends = fetch_ends[order(directions), ]
    
    ## Create a list of Line objects radiating from the site location to the 
    ## maximum distance for each direction.
    line_list = create_line_list(site_layer[i, ], fetch_ends)
    
    ## Create a SpatialLines object (i.e. add in CRS information)
    fetch_sp_lines = create_sp_lines(line_list, sort(directions), polygon_layer)
    
    ## Subset polygon coastline layer to only incorporate polygons that 
    ## interfere with any of the fetch vectors. This will speed up computation 
    ## times.
    poly_layer_subset = 
      polygon_layer[which(!is.na(over(polygon_layer, fetch_sp_lines))), ]
    
    if (length(poly_layer_subset) > 0){
      
      ## Calculate the fetch bearings that hit land
      hit_land = !sapply(gIntersects(fetch_sp_lines, poly_layer_subset,
                                     byid = c(TRUE, FALSE), returnDense = FALSE),
                         is.null)
      
      ## Calculate intersections and identify closest shoreline for those 
      ## vectors that hit land
      ints = gIntersection(fetch_sp_lines[hit_land], poly_layer_subset, 
                           byid = c(TRUE, FALSE))
      
      fetch_ends[hit_land, ] = t(sapply(ints@lines, function(x){
        coordinates(x)[[1]][1, ]
      }))
      
      ## Update the line list and spatialLines objects
      line_list = create_line_list(site_layer[i, ], fetch_ends)
      fetch_sp_lines = create_sp_lines(line_list, sort(directions), polygon_layer)
    }
    
    fetch.df = data.frame(site = site_names[i],
                          fetch = SpatialLinesLengths(fetch_sp_lines) / 1000,
                          direction = sort(directions),
                          quadrant = factor(quadrant, 
                                            levels = c("North", "East",
                                                       "South", "West")))
    
    ## Create a SpatialLinesDataFrame object to include the fetch lengths, and 
    ## add it to the Fetch list
    fetch_list[[i]] = SpatialLinesDataFrame(fetch_sp_lines, fetch.df)
  }
  new("Fetch", fetch_list, names = site_names, max_dist = max_dist / 1000)
}